src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java

```diff
@@ -14,30 +14,171 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import javax.servlet.ServletException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import org.apache.http.client.utils.URIBuilder;
 
-/**
- * Provides information on a lecture.
- */
+/** Provides information on a lecture. */
 @WebServlet("/lecture")
 public class LectureServlet extends HttpServlet {
+  /* Used to generate a Pattern for a Video URL. */
+  private static final String YOUTUBE_VIDEO_URL_PATTERN =
+      "(?<=watch\\?v=|/videos/|embed\\/|youtu.be\\/|\\/v\\/|\\/e\\/|"
+      + "watch\\?v%3D|watch\\?feature=player_embedded&v=|%2Fvideos%2F|"
+      + "embed%\u200C\u200B2F|youtu.be%2F|%2Fv%2F)[^#\\&\\?\\n]*";
+
+  /* Name of input field used for lecture name in lecture selection page. */
+  private static final String NAME_INPUT = "name-input";
+  
+  /* Name of input field used for lecture video link in lecture selection page. */
+  private static final String LINK_INPUT = "link-input";
+
+  private static final String REDIRECT_URL = "/lecture-view.html";
+
+  /* Pattern used to create a matcher for a video ID. */
+  private static Pattern videoUrlGeneratedPattern;
+
+  private static DatastoreService datastore;
+
   @Override
   public void init() throws ServletException {
-    // TODO: Implement LectureServlet.
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    videoUrlGeneratedPattern = Pattern.compile(YOUTUBE_VIDEO_URL_PATTERN);
   }
 
+  // TODO: Check if URL is valid.
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.sendError(500, "Not Implemented");
+    Optional<String> optionalVideoUrl = getParameter(request, LINK_INPUT);
+    Optional<Entity> existingEntity = checkUrlInDatabase(optionalVideoUrl);
+
+    if (existingEntity.isPresent()) {
+      response.sendRedirect(buildRedirectUrl(existingEntity.get()).get());
+      return;
+    }
+    Entity lectureEntity = createLectureEntity(request);
+    datastore.put(lectureEntity);
+    response.sendRedirect(buildRedirectUrl(lectureEntity).get());
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.sendError(500, "Not Implemented");
+    List<Lecture> lectures = getLectures();
+    Gson gson = new Gson();
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(lectures));
+  }
+
+  /**
+   * Returns the Entity in database that has {@code url}, or
+   * {@code Optional.empty()} if one doesn't exist.
+   */
+  private Optional<Entity> checkUrlInDatabase(Optional<String> videoUrl) {
+    if(!videoUrl.isPresent()) {
+      return Optional.empty();
+    }
+    String url = videoUrl.get();
```

Comment:
nit: I recommend you declare this where it is used (right before the for loop). The reason for this is it reduces cognitive load. The user does not need to keep track of this variable and what it will be used for until it is actually used.